core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -607,4 +616,32 @@ public Optional<ServiceLockData> getLockData(ServiceLockPath path) {
     return ServiceLockData.parse(lockData);
   }
 
+  @SuppressWarnings("deprecation")
+  @Override
+  public void process(WatchedEvent event) {
+    // ZooCache implements the Watcher interface so that it can
+    // receive events from the ZooSesssion session watcher.
+    switch (event.getState()) {
+      case SyncConnected:
+        log.info("{} ZooKeeper connection established, re-establishing watchers; {}", cacheId,
+            event);
+        clear();
+        setupWatchers(pathsToWatch);
+        break;
+      case Closed:
+      case Disconnected:
+        this.watchersSet.set(false);
```

Comment:
When zoocache has registered a watcher w/ multiple zookeepers then the events from different zookeepers could arrive in any order.  Like the Disconnected event from an old zookeeper could arrive after the SyncConnected event from a newer zookeeper.  Setting this false seems here seems to assume the events will arrive in an expected order.  This is the type of stuff that makes the case for having a zoocache only deal w/ a single zookeeper obj, then when an event arrives we can reason about it much more easily.